### ISSUE TITLE
Drop Python 2 support eventually

### ISSUE THREAD

**Issue by bitprophet** [paramiko/paramiko#1739]
Emphasis on _eventually_, please don't freak out (yet).

This has been on everyone's minds for years now, this particular ticket is spurred by https://github.com/pyca/cryptography/issues/5359 - being our primary dependency. Word on the street is they will drop Python 2 support in late 2020, after which if we continue to rely on prior versions, we'll be back where we were with PyCrypto re: increasingly insecure and unable to support newer key types, kexen etc (yes, even given how slow we are to add those anyways).

Did some basic number crunching using the PyPI public BigQuery corpus:

- The 12mo period starting a year ago (so 2018.08-2019.08): https://docs.google.com/spreadsheets/d/1m-hhQ4lCB_X-Ui0WZmlsRib9ZNZKe87SU4vCbBiQ0Ec/edit?usp=sharing
  - 65% Python 2.7 or unspecified
  - 35% Python 3.x
- The last 12mo (2019.08-2020.08): https://docs.google.com/spreadsheets/d/19OkI1oyY6fPRQgPWP90aa3PdmhikQb4gZE4hm0wxt4A/edit?usp=sharing
  - Basically a reversal, which is nice and presumably due in part to CPython themselves EOL'ing 2.7
  - 40% Python 2.7 or unspecified
  - 60% Python 3

40% still on 2.7 is well above my personal comfort level for leaving out in the cold, but at the same time, I only have so much time and energy these days so anything I can do to make this software easier to support, is good.

We'll see what the next 4-6mo brings.

**Comment by bitprophet** [comment]
I always forget that pypistats exists now! https://pypistats.org/packages/paramiko - they look back only about 3mo and the view from that perspective is even better, very roughly 25% Python 2 with notable downward slope.

**Comment by septatrix** [comment]
Also at this point in time most linux distributions ship Python 3 as the default and those which default to Python 2 reached their EOL for the most part. Furthermore can users fix their version to the current major version if they desire and the python2-less version should be released as a major release.

**Comment by bitprophet** [comment]
http://bitprophet.org/blog/2021/02/25/byethon2/ - I'll close this now since otherwise I'd probably forget to come back and update the status 😒 It's definitely happening!
